src/hotspot/share/gc/g1/g1ConcurrentMark.cpp

```diff
@@ -1461,71 +1461,48 @@ class G1CMDrainMarkingStackClosure : public VoidClosure {
   }
 };
 
-// Implementation of AbstractRefProcTaskExecutor for parallel
-// reference processing at the end of G1 concurrent marking
-
-class G1CMRefProcTaskExecutor : public AbstractRefProcTaskExecutor {
-private:
-  G1CollectedHeap*  _g1h;
-  G1ConcurrentMark* _cm;
-  WorkGang*         _workers;
-  uint              _active_workers;
+class G1CMRefProcClosureContext : public AbstractClosureContext {
+  uint                          _max_workers;
+  G1ConcurrentMark&             _cm;
+  G1CMIsAliveClosure            _is_alive;
+  G1CMKeepAliveAndDrainClosure* _keep_alive;
+  G1CMDrainMarkingStackClosure* _complete_gc;
+  ThreadModel                   _tm;
 
 public:
-  G1CMRefProcTaskExecutor(G1CollectedHeap* g1h,
-                          G1ConcurrentMark* cm,
-                          WorkGang* workers,
-                          uint n_workers) :
-    _g1h(g1h), _cm(cm),
-    _workers(workers), _active_workers(n_workers) { }
-
-  virtual void execute(ProcessTask& task, uint ergo_workers);
-};
-
-class G1CMRefProcTaskProxy : public AbstractGangTask {
-  typedef AbstractRefProcTaskExecutor::ProcessTask ProcessTask;
-  ProcessTask&      _proc_task;
-  G1CollectedHeap*  _g1h;
-  G1ConcurrentMark* _cm;
-
-public:
-  G1CMRefProcTaskProxy(ProcessTask& proc_task,
-                       G1CollectedHeap* g1h,
-                       G1ConcurrentMark* cm) :
-    AbstractGangTask("Process reference objects in parallel"),
-    _proc_task(proc_task), _g1h(g1h), _cm(cm) {
-    ReferenceProcessor* rp = _g1h->ref_processor_cm();
-    assert(rp->processing_is_mt(), "shouldn't be here otherwise");
-  }
-
-  virtual void work(uint worker_id) {
-    ResourceMark rm;
-    G1CMTask* task = _cm->task(worker_id);
-    G1CMIsAliveClosure g1_is_alive(_g1h);
-    G1CMKeepAliveAndDrainClosure g1_par_keep_alive(_cm, task, false /* is_serial */);
-    G1CMDrainMarkingStackClosure g1_par_drain(_cm, task, false /* is_serial */);
-
-    _proc_task.work(worker_id, g1_is_alive, g1_par_keep_alive, g1_par_drain);
-  }
+  G1CMRefProcClosureContext(
+    uint max_workers,
+    G1CollectedHeap& g1h,
+    G1ConcurrentMark& cm)
+    : _max_workers(max_workers),
+      _cm(cm),
+      _is_alive(&g1h),
+      _keep_alive(NEW_C_HEAP_ARRAY(G1CMKeepAliveAndDrainClosure, _max_workers, mtGC)),
+      _complete_gc(NEW_C_HEAP_ARRAY(G1CMDrainMarkingStackClosure, _max_workers, mtGC)),
+      _tm(ThreadModel::Single) {}
+
+  ~G1CMRefProcClosureContext() {
+    FREE_C_HEAP_ARRAY(G1CMKeepAliveAndDrainClosure, _keep_alive);
+    FREE_C_HEAP_ARRAY(G1CMDrainMarkingStackClosure, _complete_gc);
+  }
+
+  BoolObjectClosure* is_alive(uint worker_id) { return &_is_alive; }
+  OopClosure* keep_alive(uint worker_id)      {
+    assert(worker_id < _max_workers, "sanity");
+    return ::new (&_keep_alive[index(worker_id, _tm)]) G1CMKeepAliveAndDrainClosure(&_cm, _cm.task(index(worker_id, _tm)), _tm == ThreadModel::Single );
+  }
+  VoidClosure* complete_gc(uint worker_id)    {
```

Comment:
Inconsistent spacing between methods, sometimes there is a newline, sometimes not. I would prefer the additional newline. This is also the case in the other file.

Reply:
I will do that.